internal/cosesign1/misc.go

```diff
@@ -0,0 +1,175 @@
+
+	return base64Cert
+}
+
+func keyToBase64(key any) string {
+	derKey, err := x509.MarshalPKIXPublicKey(key)
+	if err != nil {
+		return ""
+	}
+	base64Key := base64.StdEncoding.EncodeToString(derKey)
+
+	return base64Key
+}
+
+/*
+	Create a pem of the form:
+
+-----BEGIN CERTIFICATE-----
+single line base64 standard encoded raw DER certificate
+-----END CERTIFICATE-----
+
+	Note that there are no extra line breaks added and that a string compare will need to accomodate that.
+*/
+
+func x509ToPEM(cert *x509.Certificate) string {
+	base64Cert := x509ToBase64(cert)
+	return base64CertToPEM(base64Cert)
+}
+
+func base64CertToPEM(base64Cert string) string {
+	var begin = "-----BEGIN CERTIFICATE-----\n"
+	var end = "\n-----END CERTIFICATE-----"
+
+	pemData := begin + base64Cert + end
+
+	return pemData
+}
+
+func keyToPEM(key any) string {
+	base64Key := keyToBase64(key)
+	return base64PublicKeyToPEM(base64Key)
+}
+
+func base64PublicKeyToPEM(base64Key string) string {
+	var begin = "-----BEGIN PUBLIC KEY-----\n"
+	var end = "\n-----END PUBLIC KEY-----"
+
+	pemData := begin + base64Key + end
+	return pemData
+}
+
+func PrintCert(name string, x509cert *x509.Certificate) {
+	logrus.Printf("%s:\n", name)
+	logrus.Printf("  Issuer = %s\n", x509cert.Issuer.String())
+	logrus.Printf("  Subject = %s\n", x509cert.Subject.String())
+	logrus.Printf("  AuthorityKeyId = %q\n", x509cert.AuthorityKeyId)
+	logrus.Printf("  SubjectKeyId = %q\n", x509cert.SubjectKeyId)
+
+	var pem = x509ToPEM(x509cert) // blob of the leaf x509 cert reformatted into pem (base64) style as per the fragment policy rules expect
+	var pubKey = x509cert.PublicKey
+	var pubKeyPem = keyToPEM(pubKey)
+
+	logrus.Printf("  Cert PEM = \n%s\n", pem)
+	logrus.Printf("  Public Key PEM = \n%s\n", pubKeyPem)
+}
+
+func PrintChain(inputFilename string) error {
+	raw := ReadBlob(inputFilename)
+	var msg cose.Sign1Message
+	err := msg.UnmarshalCBOR(raw)
+	if err != nil {
+		return err
+	}
+
+	protected := msg.Headers.Protected
+
+	// The spec says this is ordered - leaf, intermediates, root. X5Bag is unordered and woould need sorting
+	chainDER, chainPresent := protected[cose.HeaderLabelX5Chain]
+	if !chainPresent {
+		return fmt.Errorf("x5Chain missing")
+	}
+
+	chainIA := chainDER.([]interface{})
+	for _, c := range chainIA {
+		cert, err := x509.ParseCertificate(c.([]byte))
+		if err != nil {
+			return err
+		}
+		fmt.Println(x509ToPEM(cert))
+	}
+
+	return nil
+}
+
+func StringToAlgorithm(algoType string) (cose.Algorithm, error) {
+	var algo cose.Algorithm
+	var err error = nil
+
+	switch algoType {
+	case "PS256":
+		algo = cose.AlgorithmPS256
+	case "PS384":
+		algo = cose.AlgorithmPS384
+	case "PS512":
+		algo = cose.AlgorithmPS512
+	case "ES256":
+		algo = cose.AlgorithmES256
+	case "ES384":
+		algo = cose.AlgorithmES384
+	case "ES512":
+		algo = cose.AlgorithmES512
+	case "EdDSA":
+		algo = cose.AlgorithmEd25519
+	default:
+		algo = cose.AlgorithmPS384
```

Comment:
is this intended to be a default value? if that's the case, then we could just set the default when defining the variable. However, in case of an error would it make sense to return `0` for the algorithm or some other "trivial" value?

Reply:
Fixed by Maksim.